### PR TITLE
Support Cross-Origin-Resource-Sharing (CORS)

### LINKED ISSUE #1494: CORS Filter for the proxy
### Describe your problem(s)
Apply cors filter inside the proxy, based on API level.

### Describe your solution

The configuration needs to be provided within the openAPI definition using the vendor extensions. Here is a sample configuration. 

```
x-wso2-cors:
  enabled: true #default 
  accessControlAllowOrigins:
    - test.com
    - example.com
  accessControlAllowHeaders:
    - Authorization
    - Content-Type
  accessControlAllowMethods:
    - GET
    - PUT
    - POST
  accessControlAllowCredentials: true
```
if the enabled sub-property is not provided, it is considered to be true (default value). 

### How will you implement it

Needs to engage the envoy inbuilt cors filter as a HTTP filter. 
[1] https://www.envoyproxy.io/docs/envoy/v1.16.0/start/sandboxes/cors.html
[2] https://www.envoyproxy.io/docs/envoy/v1.16.0/api-v3/config/route/v3/route_components.proto#envoy-v3-api-msg-config-route-v3-corspolicy

---

### PR DESCRIPTION
**Pull request by VirajSalaka**
### Purpose
Support CORS in the gateway.

### Issues

Fixes #1494 

### Automation tests
 - Unit tests added: Yes
 - Integration tests added: Yes

### Tested environments

MacOS Mojave

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: adapter/pkg/oasparser/envoyconf/routesWithClusters.go [hunk 1]
```diff
@@ -175,11 +178,11 @@ func CreateRoutesWithClusters(mgwSwagger model.MgwSwagger, upstreamCerts []byte)
 				apiTitle, apiVersion, resource.GetPath())
 		}
 
-		routeP := createRoute(apiTitle, apiType, apiBasePath, apiVersion, endpointBasepath, resource.GetPath(), resource.GetMethod(), clusterRefProd, clusterRefSand)
+		routeP := createRoute(apiTitle, apiType, apiBasePath, apiVersion, endpointBasepath, resource.GetPath(), resource.GetMethod(), clusterRefProd, clusterRefSand, envoyCorsPolicy)
```
Comment (VirajSalaka) [inline adapter/pkg/oasparser/envoyconf/routesWithClusters.go:181]: Provide struct

File: enforcer/src/main/java/org/wso2/micro/gateway/enforcer/cors/CorsHeaderGenerator.java [hunk 2]
```diff
@@ -0,0 +1,247 @@
+///*
```
Comment (VirajSalaka) [inline enforcer/src/main/java/org/wso2/micro/gateway/enforcer/cors/CorsHeaderGenerator.java:1]: Remove unused code.

File: adapter/pkg/oasparser/model/constants.go [hunk 3]
```diff
@@ -29,6 +29,7 @@ const (
 	sandboxEndpoints    string = "x-wso2-sandbox-endpoints"
 	xWso2BasePath       string = "x-wso2-basePath"
 	xWso2Label          string = "x-wso2-label"
+	xWso2Cors           string = "x-wso2-cors"
```
Comment (VirajSalaka) [inline adapter/pkg/oasparser/model/constants.go:32]: Check the behavior when cors is configured in deployment.toml of APIM

File: enforcer/src/main/java/org/wso2/micro/gateway/enforcer/grpc/ExtAuthService.java [hunk 4]
```diff
@@ -52,21 +52,38 @@ public void check(CheckRequest request, StreamObserver<CheckResponse> responseOb
     }
 
     private CheckResponse buildResponse(ResponseObject responseObject) {
-        if (responseObject.getStatusCode() != 200) {
+        DeniedHttpResponse.Builder responseBuilder = DeniedHttpResponse.newBuilder();
+        HttpStatus status = HttpStatus.newBuilder().setCodeValue(responseObject.getStatusCode()).build();
+        if (responseObject.isDirectResponse()) {
+            // To handle options request
+            if (responseObject.getStatusCode() == 204) {
```
Comment (VirajSalaka) [inline enforcer/src/main/java/org/wso2/micro/gateway/enforcer/grpc/ExtAuthService.java:59]: Use the constant

File: test/test-integration/src/test/java/org/wso2am/micro/gw/tests/security/CorsTestCase.java [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+        Assert.assertEquals(pickHeader(responseHeaders, HttpHeaderNames.ACCESS_CONTROL_ALLOW_HEADERS.toString())
+                        .getValue().replaceAll(" ", ""), allowedHeaders);
+        Assert.assertNotNull(pickHeader(responseHeaders, ACCESS_CONTROL_ALLOW_CREDENTIALS_HEADER),
+                ACCESS_CONTROL_ALLOW_CREDENTIALS_HEADER + " header is unavailable");
+        Assert.assertTrue(Boolean.parseBoolean(pickHeader(responseHeaders,
+                ACCESS_CONTROL_ALLOW_CREDENTIALS_HEADER).getValue()));
+    }
+
+    @Test(description = "Success Scenario, with allow credentials is set to true.")
+    public void CheckCORSHeadersInSimpleResponse() throws IOException {
```
Comment (VirajSalaka) [inline test/test-integration/src/test/java/org/wso2am/micro/gw/tests/security/CorsTestCase.java:121]: testCORSHeadersInSimpleRequest

File: test/test-integration/src/test/java/org/wso2am/micro/gw/tests/util/HttpsClientRequest.java [hunk 6]
```diff
@@ -84,6 +81,11 @@ public static HttpResponse doGet(String requestUrl, Map<String, String> headers)
         }
     }
 
+    public static HttpResponse doOption(String requestUrl, Map<String, String> headers)
```
Comment (VirajSalaka) [inline test/test-integration/src/test/java/org/wso2am/micro/gw/tests/util/HttpsClientRequest.java:84]: Delete unused methods

Comment (VirajSalaka) [reply test/test-integration/src/test/java/org/wso2am/micro/gw/tests/util/HttpsClientRequest.java]: fixed via https://github.com/wso2/product-microgateway/pull/1560/commits/264d082eae3bd5a5add1b0c913ac403de24b92a8

File: test/test-integration/src/test/java/org/wso2am/micro/gw/tests/security/CorsTestCase.java (hunk 5 above)
Comment (VirajSalaka) [reply test/test-integration/src/test/java/org/wso2am/micro/gw/tests/security/CorsTestCase.java]: fixed via https://github.com/wso2/product-microgateway/pull/1560/commits/264d082eae3bd5a5add1b0c913ac403de24b92a8

File: enforcer/src/main/java/org/wso2/micro/gateway/enforcer/grpc/ExtAuthService.java (hunk 4 above)
Comment (VirajSalaka) [reply enforcer/src/main/java/org/wso2/micro/gateway/enforcer/grpc/ExtAuthService.java]: fixed via https://github.com/wso2/product-microgateway/pull/1560/commits/264d082eae3bd5a5add1b0c913ac403de24b92a8

Comment (VirajSalaka) [reply enforcer/src/main/java/org/wso2/micro/gateway/enforcer/grpc/ExtAuthService.java]: fixed via https://github.com/wso2/product-microgateway/pull/1560/commits/264d082eae3bd5a5add1b0c913ac403de24b92a8

File: adapter/pkg/oasparser/model/constants.go (hunk 3 above)
Comment (VirajSalaka) [reply adapter/pkg/oasparser/model/constants.go:32]: https://github.com/wso2/product-microgateway/issues/1566

File: enforcer/src/main/java/org/wso2/micro/gateway/enforcer/cors/CorsHeaderGenerator.java (hunk 2 above)
Comment (VirajSalaka) [reply enforcer/src/main/java/org/wso2/micro/gateway/enforcer/cors/CorsHeaderGenerator.java]: fixed via https://github.com/wso2/product-microgateway/commit/264d082eae3bd5a5add1b0c913ac403de24b92a8

File: adapter/pkg/oasparser/envoyconf/routesWithClusters.go (hunk 1 above)
Comment (VirajSalaka) [reply adapter/pkg/oasparser/envoyconf/routesWithClusters.go]: fixed via https://github.com/wso2/product-microgateway/pull/1560/commits/99050b59599b022404d9e2a73e4f4cab09df0148

File: adapter/pkg/oasparser/model/constants.go (hunk 3 above)
Comment (VirajSalaka) [reply adapter/pkg/oasparser/model/constants.go:32]: Created an additional issue to address this

**Review by AmaliMatharaarachchi — APPROVED** [review]

**Pull request merged**